tests/FSharp.Compiler.ComponentTests/Interop/StaticsInInterfaces.fs

```diff
@@ -68,4 +68,127 @@ let main _ =
+        {
+            .override method !0 class IGetNext`1<class RepeatSequence>::Next(!0)
+            ...
+        }
+    *)
+    #if !NETCOREAPP
+    [<Fact(Skip = "NET472 is unsupported runtime for this kind of test.")>]
+    #else
+    [<Fact>]
+    #endif
+    let ``F# generates valid IL for abstract static interface methods`` () =
+
+        let csharpLib = csharpBaseClass 
+
+        let fsharpSource =
+            """
+module StaticsTesting 
+open StaticsInInterfaces
+
+type MyRepeatSequence() =
+    interface IGetNext<MyRepeatSequence> with
+        static member Next(other: MyRepeatSequence) : MyRepeatSequence = other 
+"""
+        Fsx fsharpSource
+        |> withLangVersionPreview
+        |> withReferences [csharpLib]
+        |> compile
+        |> shouldSucceed
+        |> verifyIL [
+        """
+.class public abstract auto ansi sealed StaticsTesting
+extends [runtime]System.Object
+{
+  .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 07 00 00 00 00 00 ) 
+  .class auto ansi serializable nested public MyRepeatSequence
+  extends [runtime]System.Object
+  implements class [csLib]StaticsInInterfaces.IGetNext`1<class StaticsTesting/MyRepeatSequence>
+  {
+    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 03 00 00 00 00 00 ) 
+    .method public specialname rtspecialname 
+     instance void  .ctor() cil managed
+    {
+      
+      .maxstack  8
+      IL_0000:  ldarg.0
+      IL_0001:  callvirt   instance void [runtime]System.Object::.ctor()
+      IL_0006:  ldarg.0
+      IL_0007:  pop
+      IL_0008:  ret
+    } 
+
+    .method public hidebysig static class StaticsTesting/MyRepeatSequence 
+     'StaticsInInterfaces.IGetNext<StaticsTesting.MyRepeatSequence>.Next'(class StaticsTesting/MyRepeatSequence other) cil managed
+    {
+      .override  method !0 class [csLib]StaticsInInterfaces.IGetNext`1<class StaticsTesting/MyRepeatSequence>::Next(!0)
+      
+      .maxstack  8
+      IL_0000:  ldarg.0
+      IL_0001:  ret
+    } 
+
+  } 
+
+}
+        """]
+    
+#if !NETCOREAPP
+    [<Fact(Skip = "NET472 is unsupported runtime for this kind of test.")>]
+#else
+    [<Fact>]
+#endif
+    let ``F# can implement static methods declared in interfaces from C#`` () =
+
+        let csharpLib = csharpBaseClass 
+
+        let fsharpSource =
+            """
+open System
+open StaticsInInterfaces
+
+type MyRepeatSequence() =
+    interface IGetNext<MyRepeatSequence> with
+        static member Next(other: MyRepeatSequence) : MyRepeatSequence = other 
+
+[<EntryPoint>]
+let main _ =
+
+    let mutable str = MyRepeatSequence ()
+    let res = [ for i in 0..10 do
+                    yield string(str)
+                    str <- MyRepeatSequence.Next(str) ]
```

Comment:
This raises an interesting point. 

Under the name resolution rules of F#, interface implementations are not available via the implementing type, so following that principle `MyRepeatSequence.Next` shouldn't resolve. 

However, that begs the question about how we call `MyRepeatSequence.Next`.   For instance methods we don't have this problem because the user can always write 

```fsharp
     (someObject :> ISomeInterface).CallInterfaceMethod()
```

to make the interface call explicit. 
### Option A - expect the type authors to resolve this by authoring an explicitly accessible method, e.g. like this:

```fsharp
type MyRepeatSequence() =
    static member Next(other: MyRepeatSequence) =
           <the implementation>

    interface IGetNext<MyRepeatSequence> with
        static member Next(other) = MyRepeatSequence.Next(other)
...
MyRepeatSequence.Next(str)
```

Note that C# code using implicit interface impls will automatically have such a method available.  We should check if most of the math stuff in System.* will use explicit or implicit impls - we assume implicit.

### Option B - expect people to make a call via a type parameter

This can be used in conjunction in Option A

```fsharp
type MyRepeatSequence() =

    interface IGetNext<MyRepeatSequence> with
        static member Next(other) = MyRepeatSequence.Next(other)
...
let CallNext<'T when 'T : IGetNext<'T>> (str: 'T) =
    'T.Next(str)

CallNext<MyRepeatSequence>(str)
```

This second option is a bit unfortunate as it means extra obfuscating generic helper functions when there is actually no generic code around.  It also means users have no way of manually inlining such a helper function, and it makes another case where you don't get "closure under substitution" for F# generic code (which also happens for some SRTP calls, where generic helpers are also needed).

# Option C - make an exception for statics and have those be name-accessible.

However ambiguities can arise in the name resolution if several different unrelated interfaces implement that same named method - we could likely resolve those ambiguities but it does expose us to this kind of problem in a different way than currently.

### Option D - give some kind of explicit call syntax, e.g.

```fsharp
    (MyRepeatSequence :> IGenNext<MyRepeatSequence>).Next(str)
```

This is a bit ugly and undiscoverable. However it has the huge advantage of allow very precise concretization of generic code, e.g. imagine the user writes:

```
GenericMathCode<'T  when 'T : IMath<'T>> ( .... ) {
   blah //100s of lines
   'T.Sin(...)
   'T.Cos(...)
}
```
Then wants to accurately make this concrete to some specific type 
```
ConcreteMathCode ( .... ) {
   blah //100s of lines
   (Double :> IMath<'T>).Sin(...)
   (Double :> IMath<'T>).Cos(...)
}
```


### Conclusion 

For now we expect to go with Option A+B and possibly add Option D at some later point.

Reply:
Would it make sense, when there are no ambiguities (only a single member resolved with the name, out of the concrete type, in this case `MyRepeatSequence.Next`) to allow syntax which doesn't involve the upcast operator?

When there are ambiguities, the compiler can produce a neat error message asking to disambiguate, listing all the different possibilities.